packages/firestore/src/local/local_store.ts

```diff
@@ -153,6 +153,15 @@ export class LocalStore {
   /** Used to generate targetIDs for queries tracked locally. */
   private targetIdGenerator = TargetIdGenerator.forLocalStore();
 
+  /**
+   * Stores the document keys for local pending mutation batches. While
+   * primary clients will store all keys for all pending mutations, secondary
+   * clients will only store the keys for mutations that originated in these
+   * clients.
```

Comment:
It seems like all clients need to store all mutation keys, since secondary clients will still have applied the mutations from other clients locally (latency compensation) and need to revert to the remoteDocuments version when the write is no longer pending (i.e. stop latency compensating the value).

Else listeners would get stuck with documents with hasPendingWrites=true for other clients' writes, etc., right?

Reply:
I think this will all work automatically... but please correct me if I am wrong. I don't think there are any changes that need to be relayed across tabs that are not either some client's own pending mutation or part of a watch target. If a document is part of a latency compensated view, then it should also raise a Watch event in the primary tab. The primary tab will then relay these changes with the Watch change log.

Am I missing something?

Reply:
I think it's possible there's no watch update (the write didn't actually change anything) but the most obvious case is probably rejected mutations.  There definitely won't be a watch update, but all clients definitely need to stop latency-compensating the document and refresh from remoteDocuments cache.

Reply:
Wouldn't that change the metadata on the document? I need to relay both the updates from the server as well as from local mutations for all Watch targets.

Reply:
To clarify why I am a little bit concerned about this change: There is no guarantee that a client will see every pending mutation. For throttled clients, the mutation might already have been acknowledged and removed from the queue by the time it was picked up from Local Storage.

Reply:
Maybe I don't understand this watch changelog thing...  I assumed it would only reflect what we receive from watch for active targets.  It would be unaware of mutations and the local view of documents (the contents of documents with local mutations applied).

In the case of a rejected mutation (that never shows up from watch), what do you expect to go on the changelog?  Keep in mind that in order for the primary tab to even know that targets could be affected by the mutation it would need to be evaluating those queries locally, which normally it wouldn't do unless the primary tab also had an active listener (with a View, etc.) for the query.

Reply:
I moved the cache to the mutation queue for now. It definitely doesn't hurt to have it there, and the memory mutation queue's implementation is actually super straightforward.

I am still not 100% convinced that the original design doesn't work, but this is more of a matter of optimization and I much rather just look at that later. You are correct though that the primary tab won't have an active listener - and that might make it infeasible to compute the latency compensated views.